### PR TITLE
Parse nested fields in `expression_to_schema` util

### LINKED ISSUE #88: [FEATURE] Support nested types in `triad.utils.pyarrow.expression_to_schema`
**Is your feature request related to a problem? Please describe.**

Over at https://github.com/kedro-org/kedro, we've got a need to be able to specify a PyArrow schema argument in YAML. From what we can tell, PyArrow doesn't natively support defining lists/dicts/etc., so I found Fugue. However, as @avsolatorio pointed out, Fugue doesn't seem to support nested object. I think it could be pretty trivial, though?

**Describe the solution you'd like**
See #89 

For example:

```python
>>> from triad.utils.pyarrow import expression_to_schema
>>> expression_to_schema("a:[[int]]")
a: list<item: list<item: int32>>
  child 0, item: list<item: int32>
      child 0, item: int32
```

Happy to add tests; just wanted to raise the issue first to see if such a change were even welcome.

**Describe alternatives you've considered**
Rolling our own parser.

**Additional context**

> I don't think PyArrow provides a native way to define complex schema as text (as mentioned in your issue), although simple stuff like `pa.schema({'hi': 'u1'}) == pa.schema({'hi': pa.uint8()})` works. However, [Fugue](https://fugue-tutorials.readthedocs.io/tutorials/advanced/schema_dataframes.html#schema) provides a way to do so. Maybe we should just use their logic, since it's non-trivial? The requirements for https://github.com/fugue-project/triad also look pretty lightweight, so I don't see an issue with depending on it.

@deepyaman thanks for looking into this! I checked Fugue, and I don't think it supports complex schema structures such as nested list types. For example, I don't see any way to represent a field that has this representation `list_(list_(int64))`.

I have implemented a parser instead to supplant the use of `eval`, and I think it works relatively well.

_Originally posted by @avsolatorio in https://github.com/kedro-org/kedro/issues/1746#issuecomment-1201945358_

### PR DESCRIPTION
**Pull request by deepyaman**
Closes #88

### CODE REVIEW

**Comment by goodwanghan** [comment]
@deepyaman can you bring the test coverage to 100% also, please add the tests for `[[int]]` and update the file for a new release, here is a reference PR https://github.com/fugue-project/triad/pull/87/files

**Comment by goodwanghan** [comment]
Please also update https://github.com/fugue-project/triad/blob/master/triad_version/__init__.py#L2

File: triad/utils/pyarrow.py [hunk 1]
```diff
@@ -403,22 +403,23 @@ def _type_to_expression(dt: pa.DataType) -> str:
...
 
+def _parse_types(v: Any):
+    if isinstance(v, str):
+        return _parse_type(v)
+    elif isinstance(v, Dict):
+        return pa.struct(_construct_struct(v))
+    elif isinstance(v, List):
+        assert_or_throw(1 == len(v), SyntaxError(f"{v} is not a valid list type"))
+        return pa.list_(_parse_types(v[0]))
+    else:
```
Comment (goodwanghan) [inline triad/utils/pyarrow.py:414]: I suggest you add back `# pragma: no cover` to bring the coverage to 100% for now. Alternatively, you can add a test to cover the scenario.

Comment (deepyaman) [reply triad/utils/pyarrow.py]: I tried adding a test to cover the scenario, but I'm actually not sure how to reach it. 😅 

Is this expected?

```python
>>> from triad.utils.pyarrow import expression_to_schema
>>> expression_to_schema("a:1")
Traceback (most recent call last):
  File "/Users/deepyaman/triad/triad/utils/pyarrow.py", line 138, in expression_to_schema
    return pa.schema(_construct_struct(obj))
  File "pyarrow/types.pxi", line 3242, in pyarrow.lib.schema
  File "/Users/deepyaman/triad/triad/utils/pyarrow.py", line 422, in _construct_struct
    yield pa.field(k, _parse_types(v))
  File "/Users/deepyaman/triad/triad/utils/pyarrow.py", line 408, in _parse_types
    return _parse_type(v)
  File "/Users/deepyaman/triad/triad/utils/pyarrow.py", line 426, in _parse_type
    name, args = _parse_type_function(expr)
  File "/Users/deepyaman/triad/triad/utils/pyarrow.py", line 443, in _parse_type_function
    assert validate_column_name(name), f"Invalid expression {expr}"
AssertionError: Invalid expression 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/deepyaman/triad/triad/utils/pyarrow.py", line 142, in expression_to_schema
    raise SyntaxError(f"Invalid syntax: '{expr}' {str(e)}")
SyntaxError: Invalid syntax: 'a:1' Invalid expression 1
```

I'm not sure why the key (type) should be checked with `validate_column_name`.

**Review by goodwanghan — APPROVED** [review]

File: triad/utils/pyarrow.py (hunk 1 above)
Comment (goodwanghan) [reply triad/utils/pyarrow.py]: The naming of the private function needs improvement. I double checked the logic, it is correct.

**Comment by deepyaman** [comment]
@goodwanghan Thanks for the help! We'll go ahead with leveraging this for PyArrow schema parsing in [Kedro](https://github.com/kedro-org/kedro/). :)

**Comment by deepyaman** [comment]
@goodwanghan @kvnkho Thanks both for the help! We'll go ahead with leveraging this for PyArrow schema parsing in [Kedro](https://github.com/kedro-org/kedro/). :)

**Pull request merged**